contrib/bsp/src/mill/contrib/BSP.scala

```diff
@@ -0,0 +1,137 @@
+package mill.contrib
+
+import java.io.PrintWriter
+import java.nio.file.FileAlreadyExistsException
+import java.util.concurrent.Executors
+
+import ch.epfl.scala.bsp4j._
+import mill._
+import mill.define.{Command, Discover, ExternalModule}
+import mill.eval.Evaluator
+import org.eclipse.lsp4j.jsonrpc.Launcher
+import upickle.default._
+
+import scala.collection.JavaConverters._
+import scala.concurrent.CancellationException
+
+case class BspConfigJson(name: String,
+                         argv: Seq[String],
+                         version: String,
+                         bspVersion: String,
+                         languages: Seq[String])
+  extends BspConnectionDetails(name, argv.asJava, version, bspVersion, languages.asJava) {
+}
+
+object BspConfigJson {
+  implicit val rw: ReadWriter[BspConfigJson] = macroRW
+}
+
+object BSP extends ExternalModule {
+
+  implicit def millScoptEvaluatorReads[T] = new mill.main.EvaluatorScopt[T]()
+
+  lazy val millDiscover: Discover[BSP.this.type] = Discover[this.type]
+  val version = "1.0.0"
+  val bspVersion = "2.0.0"
```

Comment:
```suggestion
  /** The version of the bsp server contrib module. */
  val version = "1.0.0"
  /** The version of the BSP protocol. */
  val bspVersion = "2.0.0"
```

A different name might be even more documenting, e.g. `bspProtocolVersion`.